### PR TITLE
Add heartbeat test + Fix monitor.py

### PR DESCRIPTION
**Pull request by richardliaw**
## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by richardliaw** [comment]
@zhijunfu @raulchen @pcmoritz we're debugging the autoscaler, and it looks like the heartbeats are completely missing resource data. After a bisection, it looks like it stems from #5024; do you know what to do to fix this?

Thanks!

**Comment by raulchen** [comment]
@richardliaw sorry for introducing this bug. It should be fixed by the following patch. Do you want to include the patch in this PR? Or I can submit another fix as well.
```
diff --git a/src/ray/raylet/node_manager.cc b/src/ray/raylet/node_manager.cc
index 2b26579ae..b8ded2e6b 100644
--- a/src/ray/raylet/node_manager.cc
+++ b/src/ray/raylet/node_manager.cc
@@ -547,8 +547,8 @@ void NodeManager::HeartbeatAdded(const ClientID &client_id,
   SchedulingResources &remote_resources = it->second;
 
   ResourceSet remote_available(
-      VectorFromProtobuf(heartbeat_data.resources_total_label()),
-      VectorFromProtobuf(heartbeat_data.resources_total_capacity()));
+      VectorFromProtobuf(heartbeat_data.resources_available_label()),
+      VectorFromProtobuf(heartbeat_data.resources_available_capacity()));
   ResourceSet remote_load(VectorFromProtobuf(heartbeat_data.resource_load_label()),
                           VectorFromProtobuf(heartbeat_data.resource_load_capacity()));
   // TODO(atumanov): assert that the load is a non-empty ResourceSet.
```

**Comment by richardliaw** [comment]
Hi Hao,

Does this patch fix the test? Right now I think we are missing all resource
information when we receive a heartbeat message.

Feel free to push to this branch too.

**Comment by hartikainen** [comment]
Hey @raulchen, thanks for the patch! I tried those changes, but they don't seem to solve the underlying issue. As Richard said, it seems like all the resource information is missing currently.

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -547,8 +547,8 @@ void NodeManager::HeartbeatAdded(const ClientID &client_id,
   SchedulingResources &remote_resources = it->second;
 
   ResourceSet remote_available(
-      VectorFromProtobuf(heartbeat_data.resources_total_label()),
-      VectorFromProtobuf(heartbeat_data.resources_total_capacity()));
+      VectorFromProtobuf(heartbeat_data.resources_available_label()),
```
Comment (hartikainen) [inline src/ray/raylet/node_manager.cc]: This probably needs a test.

**Comment by richardliaw** [comment]
@raulchen @hartikainen it looks like the patch was the main CPP change that was needed. I found out the main issue was the resource_count of 0 == no resource entry. 

@raulchen can you add a test for the resource set change? Thanks!

@hartikainen can you review the autoscaler changes and try this out? I think this should work (or we have to write more comprehensive things...)

File: python/ray/monitor.py [hunk 2]
```diff
@@ -106,25 +106,23 @@ def xray_heartbeat_batch_handler(self, unused_channel, data):
...
+            avail_resources = dict(
+                zip(heartbeat_message.resources_available_label,
+                    heartbeat_message.resources_available_capacity))
+            for res in total_resources:
+                avail_resources.setdefault(res, 0.0)
 
             # Update the load metrics for this raylet.
             client_id = ray.utils.binary_to_hex(heartbeat_message.client_id)
             ip = self.raylet_id_to_ip_map.get(client_id)
+            load_metrics_id = ip + "-" + client_id
```
Comment (hartikainen) [inline python/ray/monitor.py]: Is this the best way to do this?

File: python/ray/monitor.py [hunk 3]
```diff
@@ -106,25 +106,23 @@ def xray_heartbeat_batch_handler(self, unused_channel, data):
...
-                dyn = heartbeat_message.resources_available_label[i]
-                static = heartbeat_message.resources_total_label[i]
-                dynamic_resources[dyn] = (
-                    heartbeat_message.resources_available_capacity[i])
-                static_resources[static] = (
-                    heartbeat_message.resources_total_capacity[i])
+            total_resources = dict(
+                zip(heartbeat_message.resources_total_label,
+                    heartbeat_message.resources_total_capacity))
+            avail_resources = dict(
```
Comment (hartikainen) [inline python/ray/monitor.py]:
```suggestion
            available_resources = dict(
```

File: python/ray/monitor.py [hunk 4]
```diff
@@ -106,25 +106,23 @@ def xray_heartbeat_batch_handler(self, unused_channel, data):
...
+                zip(heartbeat_message.resources_available_label,
+                    heartbeat_message.resources_available_capacity))
+            for res in total_resources:
+                avail_resources.setdefault(res, 0.0)
```
Comment (hartikainen) [inline python/ray/monitor.py]:
```suggestion
                available_resources.setdefault(resource, 0.0)
```

File: python/ray/monitor.py [hunk 5]
```diff
@@ -106,25 +106,23 @@ def xray_heartbeat_batch_handler(self, unused_channel, data):
...
-                    heartbeat_message.resources_available_capacity[i])
-                static_resources[static] = (
-                    heartbeat_message.resources_total_capacity[i])
+            total_resources = dict(
+                zip(heartbeat_message.resources_total_label,
+                    heartbeat_message.resources_total_capacity))
+            avail_resources = dict(
+                zip(heartbeat_message.resources_available_label,
+                    heartbeat_message.resources_available_capacity))
+            for res in total_resources:
```
Comment (hartikainen) [inline python/ray/monitor.py]:
```suggestion
            for resource in total_resources:
```

File: python/ray/monitor.py [hunk 6]
```diff
@@ -106,25 +106,23 @@ def xray_heartbeat_batch_handler(self, unused_channel, data):
...
 
             # Update the load metrics for this raylet.
             client_id = ray.utils.binary_to_hex(heartbeat_message.client_id)
             ip = self.raylet_id_to_ip_map.get(client_id)
+            load_metrics_id = ip + "-" + client_id
             if ip:
-                self.load_metrics.update(ip, static_resources,
-                                         dynamic_resources)
+                self.load_metrics.update(load_metrics_id, total_resources,
+                                         avail_resources)
```
Comment (hartikainen) [inline python/ray/monitor.py]:
```suggestion
                                         available_resources)
```

**Review by hartikainen — APPROVED** [review]
Looks good to me!

**Comment by raulchen** [comment]
Hi, @richardliaw, it seems that the test in this PR has already covered my bug. Should I add another one?

**Comment by hartikainen** [comment]
@raulchen, I think it would be useful to have a test for it on the backend side. What do you think?

**Comment by raulchen** [comment]
> @raulchen, I think it would be useful to have a test for it on the backend side. What do you think?

yeah, I agree that having such a test in C++ would be useful. However, it's not easy to do that. Because we don't have C++ workers. Currently, there are only some unit tests in C++ that test particular components of the backend. And integration tests are written in Python and Java.

File: src/ray/raylet/node_manager.cc (hunk 1 above)
Comment (richardliaw) [reply src/ray/raylet/node_manager.cc]: (Resolved per Hao Chen's point)

File: python/ray/monitor.py (hunk 2 above)
Comment (hartikainen) [reply python/ray/monitor.py]: I think this needs a corresponding change to somewhere else. I currently see the following in my autoscaler logs:
> 2019-07-16 17:22:38,247 INFO autoscaler.py:198 -- LoadMetrics: Removed 1 stale ip mappings: {'10.138.0.90-e7a33a624c198a4a3280ed3ee9fbe3fdd13d287e'} not in {'10.138.0.94', '10.138.0.95', '10.138.0.91', '10.138.0.92', '10.138.0.90', '10.138.0.93'}

Comment (richardliaw) [reply python/ray/monitor.py]: Hm I think this might be wrong (or requires a larger change to get right)...

**Pull request merged**